torch/_inductor/codegen/cpp.py

```diff
@@ -3801,6 +3830,47 @@ def is_lowp_fp_store(node: torch.fx.Node):
                     # Hence, we remove the first to_type.
                     to_lowp_fp_legalized_nodes.append(_node)
                     _node.args = (ops, x, torch.float)
+                elif _node.target == "to_dtype_bitcast":
+                    (ops, value_var, dtype, src_dtype) = _node.args
+
+                    # to_dtype_bitcast act as a lowp fp sink:
+                    # c10::bit_cast requires the source and target have the bitwidth. Because the input tensor's dtype
+                    # could be promoted, e.g. from float16 to float, we have to cast the tensor to its original source
+                    # dtype before invoking bit_cast.
+                    if src_dtype in DTYPE_LOWP_FP:
+                        # No need to promote to float if it is a user of a lowp fp sources
+                        # which are all directly fed to ops that accepts lowp fp input
+                        if not (
+                            is_lowp_fp_source(value_var, src_dtype)
+                            and all(
+                                is_lowp_fp_sink(user, src_dtype)
+                                for user in value_var.users
+                            )
+                        ):
+                            with sub_graph.inserting_before(_node):
+                                to_type_node = sub_graph.call_method(
+                                    "to_dtype", args=(ops, value_var, src_dtype)
+                                )
+                                _node.replace_input_with(value_var, to_type_node)
+                                metrics.cpp_to_dtype_count += 1
+
+                    # to_dtype_bitcast act as a lowp fp source:
+                    # We also need to convert the bit-casted tensor back to float to make sure we keep using higher
+                    # precision values for the rest of the computation.
+                    if dtype in DTYPE_LOWP_FP:
+                        # No need to promote to float if all users are ops that accepts lowp fp input
+                        if not (
+                            all(is_lowp_fp_sink(user, dtype) for user in _node.users)
+                        ):
+                            ops = _node.args[0]
+                            with sub_graph.inserting_after(_node):
+                                to_type_node = sub_graph.call_method(
+                                    "to_dtype", args=(ops, _node, torch.float)
+                                )
+                                to_type_node_args = to_type_node.args
```

Comment:
This line seems not used?

Reply:
There is a similar logic in the current code (for `load`)
https://github.com/pytorch/pytorch/blob/62ce3e6e84df516fdd5310d5095fa01251806f1d/torch/_inductor/codegen/cpp.py#L3734-L3746

If I understand correctly, `replace_all_uses_with` will replace the use of current `_node` also for the newly added `to_type_node`, resulting `to_type_node` converting itself. This weird logic is here to changed back the args for the `to_type_node`. 

Actually, there is a second operator `delete_user_cb` that can act as an filter. So maybe it is clearer to change it to `_node.replace_all_uses_with(to_type_node, lambda n: n is not to_type_node)`?

https://github.com/pytorch/pytorch/blob/62ce3e6e84df516fdd5310d5095fa01251806f1d/torch/fx/node.py#L714-L715

Reply:
> If I understand correctly, replace_all_uses_with will replace the use of current _node also for the newly added to_type_node, resulting to_type_node converting itself. This weird logic is here to changed back the args for the to_type_node.

ok, make sense.